[Linux] Build issue in 5.9.2/5.9.3. No speed_hi in struct ethtool_cmd

## Summary

* OS: Linux CentOS 5
* Architecture: x86_64
* Psutil version: 5.9.2/5.9.3
* Python version: Custom build 3.11.0
* Type: build issue

## Description

On CentOS 5 (yes, I know it's old) ethtool.h does not have speed_hi in struct ethtool_cmd and it also does not have the inline function ethtool_cmd_speed.

On 5.9.2 I get:
```
  gcc-4.8 -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=592 -DPSUTIL_LINUX=1 -DPSUTIL_ETHTOOL_MISSING_TYPES=1 -I/opt/python3/include/python3.11 -c psutil/_psutil_linux.c -o build/temp.linux-x86_64-cpython-311/psutil/_psutil_linux.o
  psutil/_psutil_linux.c: In function ‘psutil_net_if_duplex_speed’:
  psutil/_psutil_linux.c:444:9: warning: implicit declaration of function ‘ethtool_cmd_speed’ [-Wimplicit-function-declaration]
           uint_speed = ethtool_cmd_speed(&ethcmd);
           ^
```
On 5.9.3 I get:
```
3.11 -c psutil/_psutil_linux.c -o build/temp.linux-x86_64-cpython-311/psutil/_psutil_linux.o
  psutil/_psutil_linux.c: In function ‘psutil_ethtool_cmd_speed’:
  psutil/_psutil_linux.c:79:17: error: ‘const struct ethtool_cmd’ has no member named ‘speed_hi’
       return (ecmd->speed_hi << 16) | ecmd->speed;
```
Version 5.9.1 is fine.

Thanks

Thanks for looking into this and I agree that it is probably not worth the effort. Maybe I should finally kill CentOS 5 from my pool of test servers :-)

I did some digging and speed_hi was introduced with Linux kernel 2.6.27 : https://elixir.bootlin.com/linux/v2.6.27/source/include/linux/ethtool.h
so perhaps it should instead state in the "platforms-support-history" that Linux kernel prior to kernel 2.6.27 is not supported.

The output from psutil install:

```
pip3 install psutil==5.9.4
Collecting psutil==5.9.4
  Using cached psutil-5.9.4.tar.gz (485 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: psutil
  Building wheel for psutil (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for psutil (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [64 lines of output]
      In file included from /tmp/tmp3hkt9_w7.c:1:
      /usr/include/linux/ethtool.h:18: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:35: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:52: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:60: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:66: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:74: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:83: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:179: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:201: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:226: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:251: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:260: error: expected specifier-qualifier-list before ‘__u32’
      /usr/include/linux/ethtool.h:266: error: expected specifier-qualifier-list before ‘__u32’
      /tmp/tmp3hkt9_w7.c:1:27: warning: no newline at end of file
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-x86_64-cpython-311
      creating build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_psosx.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_psposix.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/__init__.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_common.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_psaix.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_psbsd.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_pslinux.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_pssunos.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_compat.py -> build/lib.linux-x86_64-cpython-311/psutil
      copying psutil/_pswindows.py -> build/lib.linux-x86_64-cpython-311/psutil
      creating build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_contracts.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_bsd.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_memleaks.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/__init__.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_aix.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/__main__.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_osx.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_unicode.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_process.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/runner.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_system.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_posix.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_misc.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_sunos.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_testutils.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_connections.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_linux.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      copying psutil/tests/test_windows.py -> build/lib.linux-x86_64-cpython-311/psutil/tests
      running build_ext
      building 'psutil._psutil_linux' extension
      creating build/temp.linux-x86_64-cpython-311
      creating build/temp.linux-x86_64-cpython-311/psutil
      gcc-4.8 -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=594 -DPy_LIMITED_API=0x03060000 -DPSUTIL_LINUX=1 -DPSUTIL_ETHTOOL_MISSING_TYPES=1 -I/opt/python3/include/python3.11 -c psutil/_psutil_common.c -o build/temp.linux-x86_64-cpython-311/psutil/_psutil_common.o
      gcc-4.8 -pthread -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=594 -DPy_LIMITED_API=0x03060000 -DPSUTIL_LINUX=1 -DPSUTIL_ETHTOOL_MISSING_TYPES=1 -I/opt/python3/include/python3.11 -c psutil/_psutil_linux.c -o build/temp.linux-x86_64-cpython-311/psutil/_psutil_linux.o
      psutil/_psutil_linux.c: In function ‘psutil_ethtool_cmd_speed’:
      psutil/_psutil_linux.c:84:17: error: ‘const struct ethtool_cmd’ has no member named ‘speed_hi’
           return (ecmd->speed_hi << 16) | ecmd->speed;
                       ^
      psutil/_psutil_linux.c:85:1: warning: control reaches end of non-void function [-Wreturn-type]
       }
       ^
      error: command '/opt/gcc-4.8/bin/gcc-4.8' failed with exit code 1
      [end of output]
  
```

--- a/.github/workflows/issues.py
+++ b/.github/workflows/issues.py
@@ -10,6 +10,8 @@ is created. Assign labels, provide replies, closes issues, etc. depending
 on the situation.
 """
 
+from __future__ import print_function
+
 import functools
 import json
 import os

--- a/HISTORY.rst
+++ b/HISTORY.rst
@@ -1,5 +1,10 @@
 *Bug tracker at https://github.com/giampaolo/psutil/issues*
 
+5.9.5 (IN DEVELOPMENT)
+======================
+
+- 2164_, [Linux]: compilation fails on kernels < 2.6.27 (e.g. CentOS 5).
+
 5.9.4
 =====
 

--- a/psutil/__init__.py
+++ b/psutil/__init__.py
@@ -211,7 +211,7 @@ if hasattr(_psplatform.Process, "rlimit"):
 AF_LINK = _psplatform.AF_LINK
 
 __author__ = "Giampaolo Rodola'"
-__version__ = "5.9.4"
+__version__ = "5.9.5"
 version_info = tuple([int(num) for num in __version__.split('.')])
 
 _timer = getattr(time, 'monotonic', time.time)

--- a/psutil/_psutil_linux.c
+++ b/psutil/_psutil_linux.c
@@ -78,10 +78,16 @@ ioprio_set(int which, int who, int ioprio) {
     return syscall(__NR_ioprio_set, which, who, ioprio);
 }
 
-// defined in linux/ethtool.h but not always available (e.g. Android)
+// * defined in linux/ethtool.h but not always available (e.g. Android)
+// * #ifdef check needed for old kernels, see:
+//   https://github.com/giampaolo/psutil/issues/2164
 static inline uint32_t
 psutil_ethtool_cmd_speed(const struct ethtool_cmd *ecmd) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 27)
+    return ecmd->speed;
+#else
     return (ecmd->speed_hi << 16) | ecmd->speed;
+#endif
 }
 
 #define IOPRIO_CLASS_SHIFT 13

--- a/psutil/tests/test_windows.py
+++ b/psutil/tests/test_windows.py
@@ -23,8 +23,8 @@ import warnings
 import psutil
 from psutil import WINDOWS
 from psutil._compat import FileNotFoundError
-from psutil._compat import which
 from psutil._compat import super
+from psutil._compat import which
 from psutil.tests import APPVEYOR
 from psutil.tests import GITHUB_ACTIONS
 from psutil.tests import HAS_BATTERY

